### ISSUE TITLE
Login user could not access external kube api.

### ISSUE THREAD

**Issue by xuant** [openshift/origin#3964]
## Environment:

OpenStack vms:
vm1: Etcd
vm2: Kubernetes master and node all-in-one, pointing to vm1 Etcd
vm3: Openshift master

## Kubernetes configuration

1. Starting apiserver and controller-manager according to https://github.com/GoogleCloudPlatform/kubernetes/issues/10265#issuecomment-120300549
```
api-server flags:
--logtostderr=true 
--v=0 
--etcd_config=/etc/kubernetes/etcd.json 
--address=0.0.0.0 
--allow_privileged=false 
--portal_net=10.254.0.0/16 
--admission_control=NamespaceLifecycle,NamespaceExists,LimitRanger,SecurityContextDeny,ServiceAccount,ResourceQuota 
--runtime-config=api/all=false,api/v1=true 
--client-ca-file=/etc/kubernetes/kubeca.crt 
--tls-cert-file=/etcd/kubernetes/kube.server.crt 
--tls-private-key-file=/etcd/kubernetes/kube.server.key


controller-manager flags:
--logtostderr=true 
--v=0 
--master=http://100.73.145.94:8080 
--root-ca-file=/etc/kubernetes/kubeca.crt 
--service-account-private-key-file=/etc/kubernetes/kube.server.key
```

2. Generate a kubeconfig file "external-kube.kubeconfig" using the same cert files and scp this file to the OpenShift vm :
```
$ kubectl config set-cluster/set-credentials/set-context
```

## Create OpenShift Master config

```
$ mkdir write-config
$ ./openshift start master --write-config="write-config/" --kubeconfig="external-kube.kubeconfig"
```

Then modified "write-config/master-config.yaml" by adding "kube.server.key" into ServiceAccountConfig/publicKeyFiles

## Start OpenShift master in foreground:

```
./openshift start master --config="write-config/master-config.yaml"
```

## Test
I copied "write-config/admin.kubeconfig" into ${HOME}/.kube/config so that I can operate as cluster admin. I was able to query both OpenShift api and external Kube api, but if I login as a random user, I am not able to query the external Kube api:
```
$ ./os-bin/oc get projects
NAME              DISPLAY NAME   STATUS
default                          Active
openshift                        Active
openshift-infra                  Active

$ ./os-bin/oc get nodes
NAME            LABELS                                 STATUS
100.73.145.94   kubernetes.io/hostname=100.73.145.94   Ready

$ ./os-bin/oc login
Username: test-user
Authentication required for https://100.73.145.92:8443 (openshift)
Password:
Login successful.

You don't have any projects. You can try to create a new project, by running

    $ oc new-project <projectname>

$ ./os-bin/oc new-project test-user-project
error: couldn't read version from server: User "test-user" cannot "get" on "/api/"
```

@stevekuznetsov

**Comment by stevekuznetsov** [comment]
@xuant As the `cluster:admin` user you may need to set give the appropriate role to your `test-user`. @deads2k can you weigh in?

**Comment by deads2k** [comment]
@xuant It's unusual for the `oc new-project` to try to hit `/api/` as opposed to `/api`.  Can you you include the output of `oc new-project test-3 --loglevel=6` and `oc config view`?  Remember to redact tokens and client-keys from your config output.

**Comment by xuant** [comment]
@deads2k 
After logging in as "joe" (AllowAll authentication), It first hit /api and then /api/ (I don't see it as the problem here). It is more likely some kind of role needs to be assigned to "joe":
```
./oc new-project joe-test-project --loglevel=6
I0731 18:25:02.443614   31138 debugging.go:120] GET https://100.73.145.92:8443/api 301 Moved Permanently in 20 milliseconds
I0731 18:25:02.446402   31138 debugging.go:120] GET https://100.73.145.92:8443/api/ 403 Forbidden in 2 milliseconds
F0731 18:25:02.446743   31138 helpers.go:72] error: couldn't read version from server: User "joe" cannot "get" on "/api/"
```

**Comment by deads2k** [comment]
Are you running a proxy of some sort in front of your kubernetes?  The 301 response from `/api` to `/api/` means that the default policy to allow all users the ability to negotiate an API version with the server will fail.  See https://github.com/openshift/origin/blob/master/pkg/cmd/server/bootstrappolicy/policy.go#L146.

**Comment by derekwaynecarr** [comment]
@deads2k - when we run openshift on top of kube, we are running that proxy ourselves

**Comment by deads2k** [comment]
> @deads2k - when we run openshift on top of kube, we are running that proxy ourselves

Yes, but we don't force that particular 301

**Comment by xuant** [comment]
@deads2k 
"system:admin" creates new projects successfully also containing this "301":
```
I0731 18:24:18.832208   31124 debugging.go:120] GET https://100.73.145.92:8443/api 301 Moved Permanently in 31 milliseconds
I0731 18:24:18.834014   31124 debugging.go:120] GET https://100.73.145.92:8443/api/ 200 OK in 1 milliseconds
I0731 18:24:19.360605   31124 debugging.go:120] GET https://100.73.145.92:8443/oapi/v1/projectrequests 200 OK in 85 milliseconds
I0731 18:24:20.354289   31124 debugging.go:120] POST https://100.73.145.92:8443/oapi/v1/projectrequests 201 Created in 993 milliseconds
I0731 18:24:21.139976   31124 debugging.go:120] GET https://100.73.145.92:8443/oapi/v1/projects/test-3 200 OK in 784 milliseconds
I0731 18:24:21.331462   31124 debugging.go:120] GET https://100.73.145.92:8443/oapi/v1/users/~ 200 OK in 51 milliseconds
I0731 18:24:21.538931   31124 debugging.go:120] GET https://100.73.145.92:8443/oapi/v1/users/~ 200 OK in 25 milliseconds
Now using project "test-3" on server "https://100.73.145.92:8443".
```

**Comment by deads2k** [comment]
> "system:admin" creates new projects successfully also containing this "301":

`system:admin` has the `cluster-admin` which has absolute power to every possible endpoint on the API server.  Every other role is restricted.  I've opened a pull to handle `/api/` because that endpoint now exists (didn't used to), but it would still be good to know whether you have something in front of openshift or kube or if there's some new configuration value that is issue redirects.

**Comment by liggitt** [comment]
Do you see the 301 when running against plain openshift?

**Comment by xuant** [comment]
@liggitt 
No

**Comment by xuant** [comment]
@deads2k 
Didn't put something in front of OpenShift or Kubernetes, but I believe OpenShift master starts its own proxy for the external kubernetes while starting up.

**Comment by liggitt** [comment]
It does start a proxy, but that proxy shouldn't be creating 301 responses. Is it possible the back end kubernetes has started redirecting /api to /api/?

**Comment by deads2k** [comment]
> It does start a proxy, but that proxy shouldn't be creating 301 responses. Is it possible the back end kubernetes has started redirecting /api to /api/?

I checked.  It didn't for me.

**Comment by xuant** [comment]
@liggitt @deads2k 
Not for me either.
Since `system:admin` is able to create a new project with "301" response but a normal user cannot, I am assuming the problem comes not from here whether the proxy is or isn't creating the 301 response (correct me if I'm wrong). On the other hand, the normal user "joe" receives a "403 Forbidden" response, is it more likely to be an authorization issue? If it is, shouldn't it be consistent regarding running plain OpenShift v.s. running OpenShift with external Kubernetes?

**Comment by liggitt** [comment]
The issue is the redirect. system:admin is allowed to access any URL anywhere in the system. @deads2k already submitted a PR that also allows /api/, but we're trying to figure out why the root cause (the redirect) is occurring.

**Comment by xuant** [comment]
@liggitt @deads2k I see. I;m gonna merge the PR myself and see it is works. Thanks for your help!

**Comment by deads2k** [comment]
> @liggitt @deads2k I see. I;m gonna merge the PR myself and see it is works. Thanks for your help!

When you try it, be sure to run `oadm policy reconcile-cluster-roles`

**Comment by mrGrab** [comment]
hi,
May I know PR number?
I have the same issue with external kubernetes: 

$ oc whoami --loglevel=6
I0820 09:32:47.634418    8156 debugging.go:120] GET https://[server-ip]:8443/api 301 Moved Permanently in 20 milliseconds
I0820 09:32:47.635838    8156 debugging.go:120] GET https://[server-ip]:8443/api/ 403 Forbidden in 1 milliseconds
F0820 09:32:47.636243    8156 helpers.go:71] error: couldn't read version from server: User "root" cannot "get" on "/api/"

**Comment by liggitt** [comment]
it was fixed in #3974. Make sure you update role definitions with `oadm policy reconcile-cluster-roles` if you update to that build with an existing cluster.
